pkg/api/service/warnings.go

```diff
@@ -0,0 +1,105 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"net/netip"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/core/helper"
+)
+
+func GetWarningsForService(service, oldService *api.Service) []string {
+	if service == nil {
+		return nil
+	}
+	var warnings []string
+
+	if helper.IsServiceIPSet(service) {
+		warnings = append(warnings, getWarningsForClusterIPs(service)...)
+	}
+
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		warnings = append(warnings, getWarningsForExternalName(service)...)
+	}
+
+	if oldService != nil {
+		warnings = append(warnings, getWarningsForServiceTransitions(service, oldService)...)
+	}
+
+	return warnings
+}
+
+// getWarningsForClusterIPs returns warnings for ClusterIPs
+func getWarningsForClusterIPs(service *api.Service) []string {
+	var warnings []string
+	var fieldPath *field.Path
+
+	for i, clusterIP := range service.Spec.ClusterIPs {
+		// IPs with leading zeros CVE-2021-29923 are not valid in golang since 1.17
+		// This will also warn about possible future changes on the golang std library
+		// xref: https://issues.k8s.io/108074
+		ip, err := netip.ParseAddr(clusterIP)
+		if err != nil {
+			warnings = append(warnings, fmt.Sprintf("%s: https://issues.k8s.io/108074, IP address %s invalid in a future release: %v", fieldPath.Child("spec", "clusterIPs").Index(i), clusterIP, err))
+			continue
+		}
+		// A Recommendation for IPv6 Address Text Representation
+		//
+		// "All of the above examples represent the same IPv6 address.  This
+		// flexibility has caused many problems for operators, systems
+		// engineers, and customers.
+		// ..."
+		// https://datatracker.ietf.org/doc/rfc5952/
+		if ip.Is6() && ip.String() != clusterIP {
+			warnings = append(warnings, fmt.Sprintf("%s: IP address %s not in canonical format %s (RFC 5952), invalid in a future release", fieldPath.Child("spec", "clusterIPs").Index(i), clusterIP, ip.String()))
+		}
+	}
+	return warnings
+}
+
+func getWarningsForServiceTransitions(service, oldService *api.Service) []string {
+	var warnings []string
+	var fieldPath *field.Path
+
+	if service.Spec.Type == oldService.Spec.Type {
+		return warnings
+	}
+
+	if helper.IsServiceIPSet(oldService) && service.Spec.Type == api.ServiceTypeExternalName {
+		warnings = append(warnings, fmt.Sprintf("%s: ClusterIPs %v dropped during update", fieldPath.Child("spec", "clusterIPs"), service.Spec.ClusterIPs))
+	}
+
+	return warnings
+}
+
+func getWarningsForExternalName(service *api.Service) []string {
+	var warnings []string
+	var fieldPath *field.Path
+
+	if service.Spec.Selector != nil {
+		warnings = append(warnings, fmt.Sprintf("%s: selector set on ExternalName Type Service, invalid in a future release", fieldPath.Child("spec", "selector")))
```

Comment:
```suggestion
		warnings = append(warnings, fmt.Sprintf("%s: selector found specified for a Service of type: ExternalName", fieldPath.Child("spec", "selector")))
```
or
```suggestion
		warnings = append(warnings, fmt.Sprintf("%s: selector found specified for a Service of type: ExternalName; a future Kubernetes release may stop accepting such changes", fieldPath.Child("spec", "selector")))
```
?

Reply:
How about 

```
fmt.Sprintf("%s: selector was specified with type = ExternalName; this has no effect and may be invalid in a future Kubernetes release", fieldPath.Child("spec", "selector")))
```

Though I am not sure we can ever actually do that.